[visionOS] Add delegate method to prevent docking from element fullscreen.

rsfuller opened the pull request:
#### 9338c51731f8745cb7817901be2fcc8f86e580c5
<pre>
[visionOS] Add delegate method to prevent docking from element fullscreen.
<a href="https://rdar.apple.com/145799817">rdar://145799817</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=294702">https://bugs.webkit.org/show_bug.cgi?id=294702</a>

Reviewed by Alex Christensen.

Safari on visionOS would like to prevent docking from element fullscreen when the
browser is in a certain state (spatial browsing).

To accomplish this, add a new delegate method to FullscreenClient that gives the
client the option to preventDocking (only for LinearMediaPlayer).

WKFullScreenViewController checks this method, and if configured to prevent, removes
the environment picker button / prevents it being added.

* Source/WebKit/UIProcess/API/APIFullscreenClient.h:
(API::FullscreenClient::preventDocking):
* Source/WebKit/UIProcess/API/Cocoa/_WKFullscreenDelegate.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
(WebKit::FullscreenClient::setDelegate):
(WebKit::FullscreenClient::preventDocking):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):

Canonical link: <a href="https://commits.webkit.org/296597@main">https://commits.webkit.org/296597@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c4a51fa2fbdf4e6261a192ae589f229cc9da46d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28261 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18685 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113808 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58990 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36815 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82487 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62924 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22396 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58522 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16010 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116929 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35654 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26288 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91509 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94091 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91313 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23334 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36213 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13974 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31469 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35555 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41088 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35265 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38611 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36943 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/API/APIFullscreenClient.h:
```diff
@@ -57,6 +57,10 @@ class FullscreenClient {
 #endif
+
+#if ENABLE(LINEAR_MEDIA_PLAYER)
+    virtual bool preventDocking(WebKit::WebPageProxy*) { return true; }
```
Why does this return true but the default value of FullscreenClient::preventDocking is false?

rsfuller replied:
Thanks for catching this! I flipped the naming from `allow` to `prevent` precisely because I wanted the default to be `false` not `true`, but forgot to update this when I made the change. Fixed!

rsfuller committed:
[visionOS] Add delegate method to prevent docking from element fullscreen.
rdar://145799817
https://bugs.webkit.org/show_bug.cgi?id=294702

Reviewed by Alex Christensen.

Safari on visionOS would like to prevent docking from element fullscreen when the
browser is in a certain state (spatial browsing).

To accomplish this, add a new delegate method to FullscreenClient that gives the
client the option to preventDocking (only for LinearMediaPlayer).

WKFullScreenViewController checks this method, and if configured to prevent, removes
the environment picker button / prevents it being added.

* Source/WebKit/UIProcess/API/APIFullscreenClient.h:
(API::FullscreenClient::preventDocking):
* Source/WebKit/UIProcess/API/Cocoa/_WKFullscreenDelegate.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.h:
* Source/WebKit/UIProcess/Cocoa/FullscreenClient.mm:
(WebKit::FullscreenClient::setDelegate):
(WebKit::FullscreenClient::preventDocking):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):

Canonical link: https://commits.webkit.org/296597@main
